# DW_pandas_Task2

## 20201216

DataWhale_pandas组队学习第二章，pandas基础，继续努力！

### 文件读取

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.read_table('../data/my_table_special_sep.txt', sep=' \|\|\|\| ', engine='python')

col1               col2
0   TS  This is an apple.
1   GQ    My name is Bob.
2   WT         Well done!
3   PT    May I help you?

平常使用sep='\t'这种无需用到engine='python'，不明白什么意思，拿掉看看

In [3]:
pd.read_table('../data/my_table_special_sep.txt', sep=' \|\|\|\| ')

C:\ProgramData\Anaconda3\lib\site-packages\pandas\io\parsers.py:767: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return read_csv(**locals())


col1               col2
0   TS  This is an apple.
1   GQ    My name is Bob.
2   WT         Well done!
3   PT    May I help you?

执行出来有提示，但结果好像看上去一样，百度一下解答：engine参数用于指定pandas读取文件时的解析引擎，有C和Python两种，C比较块，而Python特性比较多， 你sep使用了多个字符， C引擎是不支持的，所以会指定Python。
那应该是多个字符的原因，再试试

In [4]:
pd.read_table('../data/my_table_special_sep.txt', sep='\|\|\|\|', engine='python')

col1                 col2
0   TS    This is an apple.
1   GQ      My name is Bob.
2   WT           Well done!
3   PT      May I help you?

测试用engine='c'出错，没办法识别吧，然后把sep=' \\|\\|\\|\\| '中的空格拿掉，没啥区别啊，为了好看？

### 数据写入
一般在数据写入中，最常用的操作是把`index`设置为`False`，特别当索引没有特殊意义的时候，这样的行为能把索引在保存的时候去除。

数据写入的坑，用了一段时间没有这个设置，结果都是写好文件后打开再删除索引列

In [5]:
df_txt = pd.read_table('../data/my_table.txt')
# df_txt.to_csv('../data/my_table_special_sep.txt', sep='\t', index='False')

`pandas`中没有定义`to_table`函数，但是`to_csv`可以保存为`txt`文件，并且允许自定义分隔符，常用制表符`\t`分割：这个记住格式就行

In [6]:
!pip install tabulate

In [7]:
df_csv = pd.read_csv('../data/my_csv.csv')
print(df_csv.to_markdown())

|    |   col1 | col2   |   col3 | col4   | col5     |
|---:|-------:|:-------|-------:|:-------|:---------|
|  0 |      2 | a      |    1.4 | apple  | 2020/1/1 |
|  1 |      3 | b      |    3.4 | banana | 2020/1/2 |
|  2 |      6 | c      |    2.5 | orange | 2020/1/5 |
|  3 |      5 | d      |    3.2 | lemon  | 2020/1/7 |


In [8]:
print(df_csv.to_latex())

\begin{tabular}{lrlrll}
\toprule
{} &  col1 & col2 &  col3 &    col4 &      col5 \\
\midrule
0 &     2 &    a &   1.4 &   apple &  2020/1/1 \\
1 &     3 &    b &   3.4 &  banana &  2020/1/2 \\
2 &     6 &    c &   2.5 &  orange &  2020/1/5 \\
3 &     5 &    d &   3.2 &   lemon &  2020/1/7 \\
\bottomrule
\end{tabular}



如果想要把表格快速转换为markdown和latex语言，可以使用to_markdown和to_latex函数，此处需要安装tabulate包。

这两个知识点先留着，我两台电脑，一台tabulate包没安装，一台不知道是Anaconda自带的，还是我安装过。

### 基本函数

In [9]:
df = pd.read_csv('../data/learn_pandas.csv')
df.columns

Index(['School', 'Grade', 'Name', 'Gender', 'Height', 'Weight', 'Transfer',
       'Test_Number', 'Test_Date', 'Time_Record'],
      dtype='object')

In [10]:
df = df[df.columns[:7]]
df

School      Grade            Name  Gender  Height  \
0    Shanghai Jiao Tong University   Freshman    Gaopeng Yang  Female   158.9   
1                Peking University   Freshman  Changqiang You    Male   166.5   
2    Shanghai Jiao Tong University     Senior         Mei Sun    Male   188.9   
3                 Fudan University  Sophomore    Xiaojuan Sun  Female     NaN   
4                 Fudan University  Sophomore     Gaojuan You    Male   174.0   
..                             ...        ...             ...     ...     ...   
195               Fudan University     Junior    Xiaojuan Sun  Female   153.9   
196            Tsinghua University     Senior         Li Zhao  Female   160.9   
197  Shanghai Jiao Tong University     Senior  Chengqiang Chu  Female   153.9   
198  Shanghai Jiao Tong University     Senior   Chengmei Shen    Male   175.3   
199            Tsinghua University  Sophomore     Chunpeng Lv    Male   155.7   

     Weight Transfer  
0      46.0        N  
1      70.0        N  
2      89.0        N  
3      41.0        N  
4      74.0        N  
..      ...      ...  
195    46.0        N  
196    50.0        N  
197    45.0        N  
198    71.0        N  
199    51.0        N  

[200 rows x 7 columns]

In [11]:
df.describe()

Height      Weight
count  183.000000  189.000000
mean   163.218033   55.015873
std      8.608879   12.824294
min    145.400000   34.000000
25%    157.150000   46.000000
50%    161.900000   51.000000
75%    167.500000   65.000000
max    193.900000   89.000000

数据分析试过pandas-profiling包，分析很详细，测试过一个20万条的数据，生成的文件近70M，好大，打开时IE都死翘翘了，当源数据大的时候很费时间，不过内容真的很全，懒人必备。

In [12]:
df['School'].unique()

array(['Shanghai Jiao Tong University', 'Peking University',
       'Fudan University', 'Tsinghua University'], dtype=object)

In [13]:
df['School'].nunique()

4

In [14]:
df['School'].value_counts()

Tsinghua University              69
Shanghai Jiao Tong University    57
Fudan University                 40
Peking University                34
Name: School, dtype: int64

In [15]:
df_demo = df[['Gender','Transfer','Name']]
df_demo.drop_duplicates(['Gender','Transfer','Name'], keep=False).tail()

Gender Transfer            Name
194  Female      NaN     Yanmei Qian
196  Female        N         Li Zhao
197  Female        N  Chengqiang Chu
198    Male        N   Chengmei Shen
199    Male        N     Chunpeng Lv

In [16]:
df['Gender'].replace(['Female','Male'],[0,1]).head()

0    0
1    1
2    1
3    0
4    1
Name: Gender, dtype: int64

替换和填充应该用到比较多，数据中将对象类型转换为数字，缺失值的填充都很频繁，有一点不理解，数据分析应该是以数据对事物进行分析，主要还是依赖于数据模型，把所有不同的数据转化为可计算的数据，以此来计算出相关的预测或结果，但其特征仅是数据类特征，如何结合传统知识，融入传统分析的特征对数据进行分析？

In [17]:
s = pd.Series([-1, 1.2345, 100, -50])
s.clip(0, 2)

0    0.0000
1    1.2345
2    2.0000
3    0.0000
dtype: float64

#### 【练一练】

在 clip 中，超过边界的只能截断为边界值，如果要把超出边界的替换为自定义的值，应当如何做？

In [18]:
s.clip(0, 2).replace([0, 2],['a', 'b'])

0         a
1    1.2345
2         b
3         a
dtype: object

In [19]:
[ 'a' if i <2 else 'b' for i in s.clip(0, 2)]

['a', 'a', 'b', 'a']

先用replace替换不行，找不到小于2的方法，然后改用循环，不知道这样替换的理念是不是对的，费了九牛二虎之力Q_Q

In [20]:
df_demo = df[df.columns[:6]]
df_demo

School      Grade            Name  Gender  Height  \
0    Shanghai Jiao Tong University   Freshman    Gaopeng Yang  Female   158.9   
1                Peking University   Freshman  Changqiang You    Male   166.5   
2    Shanghai Jiao Tong University     Senior         Mei Sun    Male   188.9   
3                 Fudan University  Sophomore    Xiaojuan Sun  Female     NaN   
4                 Fudan University  Sophomore     Gaojuan You    Male   174.0   
..                             ...        ...             ...     ...     ...   
195               Fudan University     Junior    Xiaojuan Sun  Female   153.9   
196            Tsinghua University     Senior         Li Zhao  Female   160.9   
197  Shanghai Jiao Tong University     Senior  Chengqiang Chu  Female   153.9   
198  Shanghai Jiao Tong University     Senior   Chengmei Shen    Male   175.3   
199            Tsinghua University  Sophomore     Chunpeng Lv    Male   155.7   

     Weight  
0      46.0  
1      70.0  
2      89.0  
3      41.0  
4      74.0  
..      ...  
195    46.0  
196    50.0  
197    45.0  
198    71.0  
199    51.0  

[200 rows x 6 columns]

In [21]:
df_demo.sort_values(['Height'], ascending=False).head(10)

School      Grade            Name Gender  Height  \
193            Tsinghua University     Senior   Xiaoqiang Qin   Male   193.9   
2    Shanghai Jiao Tong University     Senior         Mei Sun   Male   188.9   
134  Shanghai Jiao Tong University     Senior      Gaoli Zhao   Male   186.5   
38               Peking University   Freshman       Qiang Han   Male   185.3   
23   Shanghai Jiao Tong University     Senior     Qiang Zheng   Male   183.9   
71   Shanghai Jiao Tong University  Sophomore        Feng Han   Male   183.4   
99               Peking University   Freshman  Changpeng Zhao   Male   181.3   
117  Shanghai Jiao Tong University   Freshman     Chunli Zhao   Male   180.2   
184  Shanghai Jiao Tong University   Freshman      Qiang Feng   Male   178.9   
95             Tsinghua University     Junior    Yanfeng Qian   Male   178.7   

     Weight  
193    79.0  
2      89.0  
134    83.0  
38     87.0  
23     87.0  
71     82.0  
99     83.0  
117    83.0  
184    80.0  
95     75.0

In [22]:
df_demo.sort_index(level=['Weight'], ascending=False).head()

School      Grade            Name  Gender  Height  \
199            Tsinghua University  Sophomore     Chunpeng Lv    Male   155.7   
198  Shanghai Jiao Tong University     Senior   Chengmei Shen    Male   175.3   
197  Shanghai Jiao Tong University     Senior  Chengqiang Chu  Female   153.9   
196            Tsinghua University     Senior         Li Zhao  Female   160.9   
195               Fudan University     Junior    Xiaojuan Sun  Female   153.9   

     Weight  
199    51.0  
198    71.0  
197    45.0  
196    50.0  
195    46.0

这个索引排序和我理解的不一样，我以为应该是按体重进行排序，但出来还是索引的倒序，是我用错了吗？

apply方法和窗口对象看不进去，暂时扔一边，我得赶紧完成作业去:-P

## 练习
### Ex1：口袋妖怪数据集
现有一份口袋妖怪的数据集，下面进行一些背景说明：

* `#`代表全国图鉴编号，不同行存在相同数字则表示为该妖怪的不同状态

* 妖怪具有单属性和双属性两种，对于单属性的妖怪，`Type 2`为缺失值
* `Total, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed`分别代表种族值、体力、物攻、防御、特攻、特防、速度，其中种族值为后6项之和

In [23]:
df = pd.read_csv('../data/pokemon.csv')
df.head(3)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1  2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2  3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  
0     45  
1     60  
2     80

1. 对`HP, Attack, Defense, Sp. Atk, Sp. Def, Speed`进行加总，验证是否为`Total`值。

2. 对于`#`重复的妖怪只保留第一条记录，解决以下问题：

* 求第一属性的种类数量和前三多数量对应的种类
* 求第一属性和第二属性的组合种类
* 求尚未出现过的属性组合

3. 按照下述要求，构造`Series`：

* 取出物攻，超过120的替换为`high`，不足50的替换为`low`，否则设为`mid`
* 取出第一属性，分别用`replace`和`apply`替换所有字母为大写
* 求每个妖怪六项能力的离差，即所有能力中偏离中位数最大的值，添加到`df`并从大到小排序

In [24]:
df_a = sum(df[df.columns[-6:]].sum())
df_b = sum(df[df.columns[-7]])
df_a == df_b

True

后六项累加等于Total这一项，验证为True，那就是对喽

In [25]:
df_c = df.drop_duplicates('#', keep='first')
df_c.head(10)

#        Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49       65   
1    2     Ivysaur  Grass  Poison    405  60      62       63       80   
2    3    Venusaur  Grass  Poison    525  80      82       83      100   
4    4  Charmander   Fire     NaN    309  39      52       43       60   
5    5  Charmeleon   Fire     NaN    405  58      64       58       80   
6    6   Charizard   Fire  Flying    534  78      84       78      109   
9    7    Squirtle  Water     NaN    314  44      48       65       50   
10   8   Wartortle  Water     NaN    405  59      63       80       65   
11   9   Blastoise  Water     NaN    530  79      83      100       85   
13  10    Caterpie    Bug     NaN    195  45      30       35       20   

    Sp. Def  Speed  
0        65     45  
1        80     60  
2       100     80  
4        50     65  
5        65     80  
6        85    100  
9        64     43  
10       80     58  
11      105     78  
13       20     45

In [26]:
df.shape

(800, 11)

In [27]:
df_c.shape

(721, 11)

行数减少了，重复的没了吧，按理说用法没错

In [28]:
df_c['Type 1'].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

In [29]:
df_c['Type 1'].nunique()

18

In [30]:
df_c['Type 1'].value_counts()[:3]

Water     105
Normal     93
Grass      66
Name: Type 1, dtype: int64

第一属性总共有18种类型，前三分别是Water，Normal，Grass

In [31]:
df_d = df_c.drop_duplicates(['Type 1', 'Type 2'])
df_d[['Type 1', 'Type 2']]

Type 1  Type 2
0      Grass  Poison
4       Fire     NaN
6       Fire  Flying
9      Water     NaN
13       Bug     NaN
..       ...     ...
773     Rock   Fairy
778    Ghost   Grass
790   Flying  Dragon
797  Psychic   Ghost
799     Fire   Water

[143 rows x 2 columns]

In [32]:
df_d.shape

(143, 11)

In [33]:
df_d = df_c.drop_duplicates(['Type 1', 'Type 2'], keep=False)
df_d[['Type 1', 'Type 2']].shape

(40, 2)

我原来做的第一属性和第二属性的组合种类是40种，看了视频是143种，发现我用keep=False理解错了，这个是将重复的值都去掉，那剩下是单一组合的，而默认的keep=first，是保留第一次出现的组合，那就是把所有不同组合都列出来了。

上面几题还是自己能思考出来，接下来的不行了，参照答案，继续理解学习。

In [34]:
L_all = [i+' '+j for i in df['Type 1'].unique() for j in (df['Type 1'].unique().tolist() + [' '])]
L_part = [i+' '+j for i, j in zip(df['Type 1'], df['Type 2'].replace(np.nan, ' '))]
res = set(L_all).difference(set(L_part))
len(res)

188

这个思路是将总的属性组合计算出来，再计算已有的属性组合，总的是将各列的唯一值拿出来相乘即可，已有的组合不理解，zip函数迭代组合，计算出来的值应该是有重复的啊？哦，看到set集合了，去除重复值，再用differencer返回两个集合的差异集合。

In [35]:
list_temp = [str(i).replace(str(i), 'high') if i>120 else str(i).replace(str(i), 'low') 
           if i<50 else str(i).replace(str(i), 'mid') for i in df['Attack']]
df_temp = pd.Series(data=list_temp)
df_temp.head(10)

0     low
1     mid
2     mid
3     mid
4     mid
5     mid
6     mid
7    high
8     mid
9     low
dtype: object

这个用自己的思路想出来，一看到替换就想到replace，所以还是对pandas的知识不了解，不懂得用mask。

In [36]:
list_temp = [i.replace(i, i.upper()) for i in df['Type 1']]
df_temp = pd.Series(data=list_temp)
df_temp.head()

0    GRASS
1    GRASS
2    GRASS
3    GRASS
4     FIRE
dtype: object

In [37]:
df['Type 1'].apply(lambda x:x.upper())

0        GRASS
1        GRASS
2        GRASS
3        GRASS
4         FIRE
        ...   
795       ROCK
796       ROCK
797    PSYCHIC
798    PSYCHIC
799       FIRE
Name: Type 1, Length: 800, dtype: object

In [38]:
df['Type 1'].replace({i:i.upper() for i in df['Type 1']})

0        GRASS
1        GRASS
2        GRASS
3        GRASS
4         FIRE
        ...   
795       ROCK
796       ROCK
797    PSYCHIC
798    PSYCHIC
799       FIRE
Name: Type 1, Length: 800, dtype: object

第一个replace是自己想的方法，第三个是参考答案中的方法，看来我还是习惯普通型的python处理方法。

In [100]:
df['HP'].median()

65.0

In [101]:
df[df.columns[-6:]].median()

HP         65.0
Attack     75.0
Defense    70.0
Sp. Atk    65.0
Sp. Def    70.0
Speed      65.0
dtype: float64

In [54]:
df_temp = df[df.columns[-6:]].T
df_temp.columns = [i for i in df['Name']]
df_temp

Bulbasaur  Ivysaur  Venusaur  VenusaurMega Venusaur  Charmander  \
HP              45       60        80                     80          39   
Attack          49       62        82                    100          52   
Defense         49       63        83                    123          43   
Sp. Atk         65       80       100                    122          60   
Sp. Def         65       80       100                    120          50   
Speed           45       60        80                     80          65   

         Charmeleon  Charizard  CharizardMega Charizard X  \
HP               58         78                         78   
Attack           64         84                        130   
Defense          58         78                        111   
Sp. Atk          80        109                        130   
Sp. Def          65         85                         85   
Speed            80        100                        100   

         CharizardMega Charizard Y  Squirtle  ...  Noibat  Noivern  Xerneas  \
HP                              78        44  ...      40       85      126   
Attack                         104        48  ...      30       70      131   
Defense                         78        65  ...      35       80       95   
Sp. Atk                        159        50  ...      45       97      131   
Sp. Def                        115        64  ...      40       80       98   
Speed                          100        43  ...      55      123       99   

         Yveltal  Zygarde50% Forme  Diancie  DiancieMega Diancie  \
HP           126               108       50                   50   
Attack       131               100      100                  160   
Defense       95               121      150                  110   
Sp. Atk      131                81      100                  160   
Sp. Def       98                95      150                  110   
Speed         99                95       50                  110   

         HoopaHoopa Confined  HoopaHoopa Unbound  Volcanion  
HP                        80                  80         80  
Attack                   110                 160        110  
Defense                   60                  60        120  
Sp. Atk                  150                 170        130  
Sp. Def                  130                 130         90  
Speed                     70                  80         70  

[6 rows x 800 columns]

In [102]:
df_median = df_temp.median()
df_sort = df_median.sort_values(ascending=False)
df_sort

MewtwoMega Mewtwo Y        130.0
GroudonPrimal Groudon      125.0
KyogrePrimal Kyogre        125.0
Arceus                     120.0
SalamenceMega Salamence    120.0
                           ...  
Igglybuff                   25.0
Happiny                     22.5
Magikarp                    20.0
Feebas                      20.0
Shuckle                     15.0
Length: 800, dtype: float64

In [103]:
df_median1 = df_median.reset_index()
df_median1

index      0
0                Bulbasaur   49.0
1                  Ivysaur   62.5
2                 Venusaur   82.5
3    VenusaurMega Venusaur  110.0
4               Charmander   51.0
..                     ...    ...
795                Diancie  100.0
796    DiancieMega Diancie  110.0
797    HoopaHoopa Confined   95.0
798     HoopaHoopa Unbound  105.0
799              Volcanion  100.0

[800 rows x 2 columns]

In [104]:
df_new = pd.merge(df, df_median1, left_on='Name', right_on='index')
df_new

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed                  index      0  
0         65       65     45              Bulbasaur   49.0  
1         80       80     60                Ivysaur   62.5  
2        100      100     80               Venusaur   82.5  
3        122      120     80  VenusaurMega Venusaur  110.0  
4         60       50     65             Charmander   51.0  
..       ...      ...    ...                    ...    ...  
795      100      150     50                Diancie  100.0  
796      160      110    110    DiancieMega Diancie  110.0  
797      150      130     70    HoopaHoopa Confined   95.0  
798      170      130     80     HoopaHoopa Unbound  105.0  
799      130       90     70              Volcanion  100.0  

[800 rows x 13 columns]

In [109]:
df_new['median'] = df_new[df_new.columns[5:11]].max(1)-df_new[df_new.columns[12]]
df_new['median']

0      16.0
1      17.5
2      17.5
3      13.0
4      14.0
       ... 
795    50.0
796    50.0
797    55.0
798    65.0
799    30.0
Name: median, Length: 800, dtype: float64

In [110]:
df_new.sort_values('median', ascending=False)

#                 Name   Type 1  Type 2  Total   HP  Attack  Defense  \
230  213              Shuckle      Bug    Rock    505   20      10      230   
121  113              Chansey   Normal     NaN    450  250       5        5   
261  242              Blissey   Normal     NaN    540  255      10       10   
333  306    AggronMega Aggron    Steel     NaN    630   70     140      230   
224  208  SteelixMega Steelix    Steel  Ground    610   75     125      230   
..   ...                  ...      ...     ...    ...  ...     ...      ...   
552  493               Arceus   Normal     NaN    720  120     120      120   
760  690               Skrelp   Poison   Water    320   50      60       60   
538  481              Mesprit  Psychic     NaN    580   80     105      105   
547  489               Phione    Water     NaN    480   80      80       80   
165  151                  Mew  Psychic     NaN    600  100     100      100   

     Sp. Atk  Sp. Def  Speed                index      0  median  
230       10      230      5              Shuckle   15.0   215.0  
121       35      105     50              Chansey   42.5   207.5  
261       75      135     55              Blissey   65.0   190.0  
333       60       80     50    AggronMega Aggron   75.0   155.0  
224       55       95     30  SteelixMega Steelix   85.0   145.0  
..       ...      ...    ...                  ...    ...     ...  
552      120      120    120               Arceus  120.0     0.0  
760       60       60     30               Skrelp   60.0     0.0  
538      105      105     80              Mesprit  105.0     0.0  
547       80       80     80               Phione   80.0     0.0  
165      100      100    100                  Mew  100.0     0.0  

[800 rows x 14 columns]

花了很长时间做了这道题，百度了N多内容，做得很复杂，然后去看看参考答案，再给自己一个大大惊叹号，答案只有短短两行，哈哈。

### Ex2：指数加权窗口
1. 作为扩张窗口的`ewm`窗口

在扩张窗口中，用户可以使用各类函数进行历史的累计指标统计，但这些内置的统计函数往往把窗口中的所有元素赋予了同样的权重。事实上，可以给出不同的权重来赋给窗口中的元素，指数加权窗口就是这样一种特殊的扩张窗口。

其中，最重要的参数是`alpha`，它决定了默认情况下的窗口权重为$w_i=(1−\alpha)^i,i\in\{0,1,...,t\}$，其中$i=t$表示当前元素，$i=0$表示序列的第一个元素。

从权重公式可以看出，离开当前值越远则权重越小，若记原序列为$x$，更新后的当前元素为$y_t$，此时通过加权公式归一化后可知：

$$
\begin{split}y_t &=\frac{\sum_{i=0}^{t} w_i x_{t-i}}{\sum_{i=0}^{t} w_i} \\
&=\frac{x_t + (1 - \alpha)x_{t-1} + (1 - \alpha)^2 x_{t-2} + ...
+ (1 - \alpha)^{t} x_{0}}{1 + (1 - \alpha) + (1 - \alpha)^2 + ...
+ (1 - \alpha)^{t-1}}\\\end{split}
$$

对于`Series`而言，可以用`ewm`对象如下计算指数平滑后的序列：

In [78]:
np.random.seed(0)
s = pd.Series(np.random.randint(-1,2,30).cumsum())
s.head()

0   -1
1   -1
2   -2
3   -2
4   -2
dtype: int32

In [79]:
s.ewm(alpha=0.2).mean().head()

0   -1.000000
1   -1.000000
2   -1.409836
3   -1.609756
4   -1.725845
dtype: float64

请用`expanding`窗口实现。

2. 作为滑动窗口的`ewm`窗口

从第1问中可以看到，`ewm`作为一种扩张窗口的特例，只能从序列的第一个元素开始加权。现在希望给定一个限制窗口`n`，只对包含自身最近的`n`个窗口进行滑动加权平滑。请根据滑窗函数，给出新的`wi`与`yt`的更新公式，并通过`rolling`窗口实现这一功能。

In [111]:
np.random.seed(0)
s = pd.Series(np.random.randint(-1,2,30).cumsum())
s.ewm(alpha=0.2).mean().head()
def ewm_func(x, alpha=0.2):
    win = (1-alpha)**np.arange(x.shape[0])[::-1]
    res = (win*x).sum()/win.sum()
    return res
s.expanding().apply(ewm_func).head()

0   -1.000000
1   -1.000000
2   -1.409836
3   -1.609756
4   -1.725845
dtype: float64

In [112]:
s.rolling(window=4).apply(ewm_func).head() # 无需对原函数改动

0         NaN
1         NaN
2         NaN
3   -1.609756
4   -1.826558
dtype: float64

#### 暂且先这样吧……
啊，撸不动了，明天有事要出门，回来得过了交作业时间，实在没办法解决第二题，大致瞄了下，一看到公式就头疼，完全懵了，只能先把作业交了，apply方法做作业时学了一些，涉及窗口对象的内容还得慢慢琢磨，不过挺好，感觉又吸收了不少知识，发现自己缺基础数学知识，还有缺时间啊。